src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -0,0 +1,398 @@
+      m_leader = true;
+      m_cond.Signal();
+    });
+
+  ctx = new FunctionContext(
+    [this, status_watcher, ctx](int r) {
+      if (r < 0) {
+        derr << "error notifying leader lock acquired: " << cpp_strerror(r)
+             << dendl;
+        delete status_watcher;
+        return;
+      }
+
+      status_watcher->init(ctx);
+    });
+
+  ctx = new FunctionContext(
+    [this, ctx](int r) {
+      Mutex::Locker locker(m_lock);
+      notify_lock_acquired(ctx);
+    });
+
+  m_work_queue->queue(ctx, 0);
+}
+
+void LeaderWatcher::release_leader_lock(Context *on_finish) {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+  assert(m_status_watcher);
+
+  MirrorStatusWatcher *status_watcher = m_status_watcher.release();
+  m_leader = false;
+  m_leader_last_heartbeat = ceph_clock_now();
+  m_cond.Signal();
+
+  on_finish = new FunctionContext(
+    [this, status_watcher, on_finish](int r) {
+      if (r < 0) {
+        derr << "error shutting mirror status watcher down: " << cpp_strerror(r)
+             << dendl;
+      }
+      Mutex::Locker locker(m_lock);
+      delete status_watcher;
+      on_finish->complete(r);
+    });
+
+  on_finish = new FunctionContext(
+    [this, status_watcher, on_finish](int r) {
+      if (r < 0) {
+        derr << "error notifying leader lock released: " << cpp_strerror(r)
+             << dendl;
+      }
+      status_watcher->shut_down(on_finish);
+    });
+
+  on_finish = new FunctionContext(
+    [this, on_finish](int r) {
+      if (r < 0) {
+        derr << "error releasing leader lock: " << cpp_strerror(r) << dendl;
+      }
+      notify_lock_released(on_finish);
+    });
+
+  m_leader_lock->release_lock(on_finish);
+}
+
+void LeaderWatcher::notify_heartbeat(Context *on_finish) {
+  dout(20) << dendl;
+
+  bufferlist bl;
+  ::encode(NotifyMessage{HeartbeatPayload{}}, bl);
+
+  librados::AioCompletion *comp = create_rados_ack_callback(on_finish);
+  int r = m_ioctx.aio_notify(m_oid, comp, bl, NOTIFY_TIMEOUT_MS, nullptr);
+  assert(r == 0);
+  comp->release();
+}
+
+void LeaderWatcher::notify_lock_acquired(Context *on_finish) {
+  dout(20) << dendl;
+
+  bufferlist bl;
+  ::encode(NotifyMessage{LockAcquiredPayload{}}, bl);
+
+  librados::AioCompletion *comp = create_rados_ack_callback(on_finish);
+  int r = m_ioctx.aio_notify(m_oid, comp, bl, NOTIFY_TIMEOUT_MS, nullptr);
+  assert(r == 0);
+  comp->release();
+}
+
+void LeaderWatcher::notify_lock_released(Context *on_finish) {
+  dout(20) << dendl;
+
+  bufferlist bl;
+  ::encode(NotifyMessage{LockReleasedPayload{}}, bl);
+
+  librados::AioCompletion *comp = create_rados_ack_callback(on_finish);
+  int r = m_ioctx.aio_notify(m_oid, comp, bl, NOTIFY_TIMEOUT_MS, nullptr);
+  assert(r == 0);
+  comp->release();
+}
+
+void LeaderWatcher::handle_heartbeat(Context *on_notify_ack) {
+  dout(20) << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    if (m_leader) {
+      derr << "got another leader heartbeat" << dendl;
```

Comment:
Seems racy -- you have filtered out that it didn't come from you (assuming you haven't had a re-watch) but due to replays it's possible that you are no longer the lock owner and you just need to handle it appropriately.